services/github-service/src/main/java/org/rhd/katapult/github/GithubResource.java

```diff
@@ -28,67 +31,119 @@
-      url.append("&client_id=").append(CLIENT_ID);
-      return Response.temporaryRedirect(URI.create(url.toString())).build();
+      log.info("Request to create repository: "+repo);
+      // Create the ghithub authorize request
+      Response response = startOauthWebFlow(repo, NEW_ACTION);
+      log.info("OAuth2 response: "+response.getStatusInfo());
+      if(log.isLoggable(Level.FINEST)) {
+         log.finest("OAuth2 response headers:");
+         response.getHeaders().forEach((key, values) -> log.info(key + ": " + values));
+      }
+      return response;
    }
 
+   /**
+    * The callback entry point used by github during the OAuth webflow to notify us that the user has approved the
+    * fork/new repository action. This is where the repository action is performed.
+    *
+    * @param code - a temporary code in a code parameter from github
+    * @param state - the state we provided when starting the OAuth webflow
+    * @return the action result reponse
+    */
    @GET
    @Path("/callback")
    public Response callback(@QueryParam("code") String code, @QueryParam("state") String state)
    {
+      log.info(String.format("/callback: code=%s, state=%s", code, state));
       String uri = "http://github.com";
       Client client = ClientBuilder.newClient();
       try
       {
-         String json = new String(Base64.getDecoder().decode(state.getBytes()));
-         JsonObject obj = Json.createReader(new StringReader(json)).readObject();
+         // Verify the state and get the repository and action to perform
+         JsonObject obj = unserializeState(state);
          String repository = obj.getString("repository");
-         JsonObject tokenObj = postToken(code, repository, client);
-         String accessToken = tokenObj.getString("access_token");
          String action = obj.getString("action");
-         if ("new".equals(action))
-         {
-            createRepository(accessToken, repository, client);
+         if(repository == null || action == null) {
+            String msg = "No repository and/or action provided in callback state: " + state;
+            log.warning(msg);
+            return Response.serverError().entity(msg).build();
          }
-         else if ("fork".equals(action))
-         {
-            JsonObject response = forkRepository(accessToken, repository, client);
-            uri = response.getString("html_url");
+
+         // Exchange the code for an access token
+         JsonObject tokenObj = postToken(code, repository, client);
+         String accessToken = tokenObj.getString("access_token");
+         // Perform the action
+         switch (action) {
+            case NEW_ACTION:
+               createRepository(accessToken, repository, client);
+               break;
+            case FORK_ACTION:
+               JsonObject response = forkRepository(accessToken, repository, client);
+               uri = response.getString("html_url");
```

Comment:
The response from the fork call is the metadata for the new repository, e.g.,

```
[
  {
    "id": 1296269,
    "owner": {
      "login": "octocat",
      "id": 1,
      "avatar_url": "https://github.com/images/error/octocat_happy.gif",
      "gravatar_id": "",
      "url": "https://api.github.com/users/octocat",
      "html_url": "https://github.com/octocat",
      "followers_url": "https://api.github.com/users/octocat/followers",
      "following_url": "https://api.github.com/users/octocat/following{/other_user}",
      "gists_url": "https://api.github.com/users/octocat/gists{/gist_id}",
      "starred_url": "https://api.github.com/users/octocat/starred{/owner}{/repo}",
      "subscriptions_url": "https://api.github.com/users/octocat/subscriptions",
      "organizations_url": "https://api.github.com/users/octocat/orgs",
      "repos_url": "https://api.github.com/users/octocat/repos",
      "events_url": "https://api.github.com/users/octocat/events{/privacy}",
      "received_events_url": "https://api.github.com/users/octocat/received_events",
      "type": "User",
      "site_admin": false
    },

```

so we are pulling out the location of the new repo page and going to that. We can certainly go back to the originating page that caused the fork request.